joomla/joomla-cms
# Improvements for secure registration and login

Assume the following setup:
- the webserver can be reached via HTTP and HTTPS
- the backend is always accessed via HTTPS ("`force_ssl`" is set to "1")
- the frontend shall be accessed by default via HTTP
- optionally the ports used for HTTP and HTTPS are non-standard

This pull request aims at the following:
- there should be a single place for activating secure login and registration
- user registration data shall be transferred via HTTPS to the server if activated
- user login at the frontend shall always use HTTPS if activated
- all pages shown when logged in at the frontend shall always be transferred via HTTPS if activated
- user logout at the frontend shall always return to HTTP if secure login is activated
- site preview from the backend shall always open the frontend in HTTP mode
- switching between HTTP and HTTPS shall work even when non-standard ports are used

Currently secure login can be activated in each `mod_login` instance separately (and there are at least three of them in the demo site delivered with Joomla), and there is no way to enable secure login when using the login form menu item coming from com_users. Moreover it is not possible to activate secure transfer for user registration data which include sensitive data like password and mail address. Also the reminding and resetting functions shall be able to use HTTPS for transferring e.g. the mail address.

In order to achieve this I moved the "usesecure" setting from the `mod_login` instance settings to the `com_users` configuration, and I changed all affected files (including message files) accordingly. In order to make foreign templates that override the `mod_login` login form work as before (assuming they check for the "usesecure" setting when logging in) I copy the "usesecure" setting from the `com_users` configuration dynamically to the `mod_login` instance settings prior to activating the login form of the template.

For proper links from the backend running in HTTPS mode to the site (site preview) I modified the administrator templates hathor and isis accordingly, moreover I had to change the administrator module `mod_status` for that purpose.

In order to handle internal URL checking even in the case of switching between HTTP and HTTPS I had to change the JUri::isInternal() method.

If you enable "Force SSL" by setting it to either "Administrator Only" or "Entire Site" (Global Configuration, Server Settings) you'll now get two additional fields for the HTTP and the HTTPS ports where you can enter the ports used at your site. This is necessary only when using non-standard ports like 8080 for HTTP and 8443 for HTTPS.

At all code locations where switching between HTTP and HTTPS is performed these ports if set are used accordingly.

Since the login and logout forms in `mod_login` use the current URL as their target for the forms and since `JRoute::_()` doesn't work properly when called with complete URLs I had to replace the calls to this method by some local code that does the switching to HTTPS if configured and the encoding of the resulting URL.

It would be easier to revert to passing "index.php" to `JRoute::_()` as it was some time ago (and as it is in all templates I've seen that contain mod_login overrides) since the target of the form doesn't matter: the real target of the login or logout process is passed in the hidden field "return". When deciding so the local code can be removed and `JRoute::_()` can be called as it is the case in the beez3 template.

I saw that the beez3 template contains an override for the `mod_login` login form but no separate override for the logout form. Instead the login form still contained the login and logout cases as was the case in `mod_login` until more than three years ago (August 2012).

Testing instructions:
- set up current Joomla version including demo site and apply all changes from this PR
- the webserver needs to be reachable via HTTP and HTTPS
- log in to the backend as administrator
- set "Force SSL" in Global Configuration, Server Settings to "Administrator Only"
- the backend should now be accessed via HTTPS only
- all site preview links should point to HTTP (hathor or isis template assumed)
- frontend login (regardless which) continue in HTTP mode
- go to the Users Configuration and enable "Encrypt Login / Registration Forms"
- check registration form: the action URL should point to HTTPS
- after successful registration the user will be back in HTTP mode
- check frontend login form (regardless which): the action URL should point to HTTPS
- after successful login all pages should be displayed via HTTPS
- in case of a login failure the user should be back in HTTP mode
- check frontend logout form: the action URL should point to HTTPS
- perform logout: the user should be back in HTTP mode
- check remind and reset links: the action URL for both of them should point to HTTPS
- after performing remind / reset functions the user should be back in HTTP mode
- change the ports used for HTTP and HTTPS, e.g. to 8080 and 8443
- configure these ports in Global Configuration, Server Settings (below "Force SSL")
- check that switching between HTTP and HTTPS works as expected
- check that the action URLs mentioned above contain the right HTTPS port

Notes:
- this PR supersedes #8003 
- this PR should be applied together with or after #8604
